src/number_theory/legendre_symbol/jacobi_symbol.lean

```diff
@@ -0,0 +1,532 @@
+lemma legendre_sym.to_jacobi_sym {p : ℕ} [fp : fact p.prime] {a : ℤ} :
+  legendre_sym p a = [a | p]ⱼ :=
+by simp only [jacobi_sym, factors_prime fp.1, list.prod_cons, list.prod_nil, mul_one, list.pmap]
+
+/-- The Jacobi symbol is multiplicative in its second argument. -/
+lemma jacobi_sym_mul_right (a : ℤ) (b₁ b₂ : ℕ) [ne_zero b₁] [ne_zero b₂] :
+  [a | b₁ * b₂]ⱼ = [a | b₁]ⱼ * [a | b₂]ⱼ :=
+begin
+  simp_rw [jacobi_sym],
+  have h := λ p hp, (list.mem_append.mp hp).elim prime_of_mem_factors prime_of_mem_factors,
+  rwa [list.perm.prod_eq (list.perm.pmap _ (perm_factors_mul (ne_zero.ne b₁) (ne_zero.ne b₂))),
+       list.pmap_append, list.prod_append],
+end
+
+/-- The Jacobi symbol `(1 / b)` has the value `1`. -/
+lemma jacobi_sym_one_left (b : ℕ) : [1 | b]ⱼ = 1 :=
+begin
+  refine rec_on_mul (jacobi_sym_zero_right 1) (jacobi_sym_one_right 1)
+                    (λ p pp, _) (λ m n hm hn, _) b,
+  { simp_rw [← @legendre_sym.to_jacobi_sym p ⟨pp⟩, @legendre_sym_one p ⟨pp⟩], },
+  { by_cases hm0 : m = 0,
+    { rw [hm0, zero_mul, jacobi_sym_zero_right], },
+    by_cases hn0 : n = 0,
+    { rw [hn0, mul_zero, jacobi_sym_zero_right], },
+    rw [@jacobi_sym_mul_right _ _ _ ⟨hm0⟩ ⟨hn0⟩, hm, hn, one_mul], },
+end
+
+/-- The Jacobi symbol is multiplicative in its first argument. -/
+lemma jacobi_sym_mul_left (a₁ a₂ : ℤ) (b : ℕ) : [a₁ * a₂ | b]ⱼ = [a₁ | b]ⱼ * [a₂ | b]ⱼ :=
+begin
+  have h0 : [a₁ * a₂ | 0]ⱼ = [a₁ | 0]ⱼ * [a₂ | 0]ⱼ :=
+  by simp only [jacobi_sym, factors_zero, list.prod_nil, one_mul, list.pmap],
+  refine rec_on_mul h0 _ (λ p pp, _) (λ m n hm hn, _) b,
+  { simp only [jacobi_sym, factors_one, list.prod_nil, one_mul, list.pmap], },
+  { simp_rw [← @legendre_sym.to_jacobi_sym p ⟨pp⟩, @legendre_sym_mul p ⟨pp⟩], },
+  { by_cases hmz : m = 0,
+    { rw [hmz, zero_mul], exact h0, },
+    by_cases hnz : n = 0,
+    { rw [hnz, mul_zero], exact h0, },
+    simp_rw [@jacobi_sym_mul_right _ _ _ ⟨hmz⟩ ⟨hnz⟩],
+    rw [hm, hn, mul_mul_mul_comm], },
+end
+
+/-- We have that `(a^e / b) = (a / b)^e` for the Jacobi symbol. -/
+lemma jacobi_sym_pow_left (a : ℤ) (e b : ℕ) : [a ^ e | b]ⱼ = [a | b]ⱼ ^ e :=
+begin
+  induction e with e ih,
+  { rw [pow_zero, pow_zero, jacobi_sym_one_left], },
+  { rw [pow_succ, pow_succ, jacobi_sym_mul_left, ih], }
+end
+
+/-- We have that `(a / b^e) = (a / b)^e` for the Jacobi symbol. -/
+lemma jacobi_sym_pow_right (a : ℤ) (b e : ℕ) : [a | b ^ e]ⱼ = [a | b]ⱼ ^ e :=
+begin
+  induction e with e ih,
+  { rw [pow_zero, pow_zero, jacobi_sym_one_right], },
+  { by_cases hb : b = 0,
+    { rw [hb, zero_pow (succ_pos e), jacobi_sym_zero_right, one_pow], },
+    { haveI : ne_zero b := ⟨hb⟩,
+      haveI : ne_zero (b ^ e) := ⟨pow_ne_zero e hb⟩,
+      rw [pow_succ, pow_succ, jacobi_sym_mul_right, ih], } }
+end
+
+/-- The Jacobi symbol `(a / b)` takes the value `1` or `-1` if `a` and `b` are coprime. -/
+lemma jacobi_sym_eq_one_or_neg_one {a : ℤ} {b : ℕ} (h : a.gcd b = 1) :
+  [a | b]ⱼ = 1 ∨ [a | b]ⱼ = -1 :=
```

Comment:
I think it's better to show the value is in the submonoid {0,1,-1} in general (using [list_prod_mem](https://leanprover-community.github.io/mathlib_docs/group_theory/submonoid/membership.html#list_prod_mem)) and then show it's nonzero when `a.gcd b = 1`. I have a proof of the latter fact that can probably be golfed:
```
lemma jacobi_sym_ne_zero {a : ℤ} {b : ℕ} (h : a.gcd b = 1) : [a | b]ⱼ ≠ 0 :=
list.prod_ne_zero begin
  simp_rw [list.mem_pmap, legendre_sym_eq_zero_iff, int_coe_zmod_eq_zero_iff_dvd],
  contrapose h,
  rw [int.gcd_eq_nat_abs, prime.not_coprime_iff_dvd],
  obtain ⟨p, hp, hd⟩ := not_not.1 h,
  exact ⟨p, prime_of_mem_factors hp, int.dvd_nat_abs_of_of_nat_dvd hd,
    int.dvd_nat_abs_of_of_nat_dvd (int.coe_nat_dvd.2 $ dvd_of_mem_factors hp)⟩,
end
```

Reply:
Now I think it's better to prove iff in one stretch:
```
theorem int.dvd_nat_abs_iff_of_nat_dvd {a : ℕ} {z : ℤ} : a ∣ z.nat_abs ↔ (a : ℤ) ∣ z :=
⟨int.of_nat_dvd_of_dvd_nat_abs, int.dvd_nat_abs_of_of_nat_dvd⟩

lemma jacobi_sym_eq_zero_iff_not_coprime {a : ℤ} {b : ℕ} [ne_zero b] :
  [a | b]ⱼ = 0 ↔ a.gcd b ≠ 1 :=
list.prod_eq_zero_iff.trans begin
  rw [list.mem_pmap, int.gcd_eq_nat_abs, ne, prime.not_coprime_iff_dvd],
  simp_rw [legendre_sym_eq_zero_iff, int_coe_zmod_eq_zero_iff_dvd, mem_factors (ne_zero.ne b),
    int.dvd_nat_abs_iff_of_nat_dvd, int.coe_nat_dvd, exists_prop, and_assoc, and_comm],
end
```
but the `b = 0` case would then need special treatment.

Reply:
OK.